src/Init/SimpLemmas.lean

```diff
@@ -45,9 +45,13 @@ theorem implies_dep_congr_ctx {p₁ p₂ q₁ : Prop} (h₁ : p₁ = p₂) {q₂
 theorem implies_congr_ctx {p₁ p₂ q₁ q₂ : Prop} (h₁ : p₁ = p₂) (h₂ : p₂ → q₁ = q₂) : (p₁ → q₁) = (p₂ → q₂) :=
   implies_dep_congr_ctx h₁ h₂
 
+-- TODO: Remove the following theorem after the next stage0 update
```

Comment:
Do we need a deprecation cycle? Is this used downstream?

Reply:
Yes, it's used in mathlib!

Reply:
It's used a few times in mathlib (my count is four files), in three of these it appears that the proof is _simpler_ if `forall_congr'` was used instead. Particularly considering the sequence of lemmas after `forall_congr'`, I believe the best final outcome is for the existing `forall_congr` to be _replaced_ by `forall_congr'` (and I have no opinion on how the deprecation / edit process should go, or how this should be actually executed).

Reply:
So what else needs to be done to initiate a deprecation cycle besides adding a `@[deprecated]` attribute?
Is this PR good to go or should I rebase on `nightly-with-mathlib`?

Reply:
> as specialised as `forall_congr'` (which IIRC works at `Prop -> Prop`).

The only difference between `forall_congr` and `forall_congr'` is that the former uses `Eq` while the latter uses `Iff`, in both input and output: the `p` and `q` inputs are at the same level of generality. Note that there's also `exists_congr`, in fact there's a series of ten lemmas starting with this one https://leanprover-community.github.io/mathlib4_docs/Init/PropLemmas.html#forall_congr', all with the same basic formulation but with a single naming discrepancy in `forall_congr'`

Reply:
Ah, of course, that makes sense. Then that still leaves `forall_congr` an instance of `pi_congr`. Deprecating the former to free up the name for `forall_congr'` seems to be The Right Thing.